File: client/src/scenes/Main/manuscriptUtils.js
import sortOn from 'sort-on';

import { groupByMultiple } from '../../utils';

const moveToFront = (list, item) => {
  if (list[0] == item) {
    // already at the front
    return list;
  }
  return [item].concat(
    list.filter(x => x !== item)
  );
};

const moveSelectPreferredManuscriptToFront = manuscripts => {
  if (manuscripts.length == 1) {
    return manuscripts;
  }
  const manuscriptsWithAbstract = manuscripts.filter(manuscript => !!manuscript.abstract);
  if (manuscriptsWithAbstract.length == 0) {
    return manuscripts;
  }
  const preferredManuscript = manuscriptsWithAbstract[0];
  return moveToFront(manuscripts, preferredManuscript);
};

export const duplicateManuscriptTitlesAsAlternatives = manuscripts => {
  if (!manuscripts || manuscripts.length < 2) {
    return manuscripts;
  }
  const manuscriptsByTitle = groupByMultiple(
    manuscripts.filter(manuscript => manuscript.title),
    manuscript => manuscript.title
  );
  Object.keys(manuscriptsByTitle).forEach(title => {
    manuscriptsByTitle[title] = moveSelectPreferredManuscriptToFront(
      manuscriptsByTitle[title]
    );
  });
  return manuscripts.map(manuscript => {
    const manuscriptsWithSameTitle = manuscriptsByTitle[manuscript.title];
    if (!manuscriptsWithSameTitle || manuscriptsWithSameTitle.length == 1) {
      return manuscript;
    }
    const preferredManuscript = manuscriptsWithSameTitle[0];
    if (manuscript == preferredManuscript) {
      return {
        ...manuscript,
        alternatives: manuscriptsWithSameTitle.slice(1)
      }
    } else {
      return null;
    }
  }).filter(manuscript => !!manuscript);
};

export const sortManuscriptsByPublishedTimestampDescending = manuscripts => manuscripts && sortOn(
  manuscripts,
  ['-published_timestamp', 'title']
);

File: peerscout/server/services/manuscript_utils.py
from itertools import groupby


def _select_preferred_manuscript_and_alternatives(manuscripts):
    if len(manuscripts) == 1:
        return manuscripts[0]
    preferred_manuscript = next(
        (m for m in manuscripts if m.get('abstract')),
        manuscripts[0]
    )
    preferred_manuscript_id = preferred_manuscript['manuscript_id']
    return {
        **preferred_manuscript,
        'alternatives': [m for m in manuscripts if m['manuscript_id'] != preferred_manuscript_id]
    }


def duplicate_manuscript_titles_as_alternatives(manuscripts):
    if not manuscripts or len(manuscripts) < 2:
        return manuscripts

    def get_title(m):
        return m.get('title', '')
    manuscripts_by_title = groupby(
        sorted(manuscripts, key=get_title),
        get_title
    )
    return [
        _select_preferred_manuscript_and_alternatives(list(manuscripts_with_same_title))
        for _, manuscripts_with_same_title in manuscripts_by_title
    ]
